File: modules/expo-pose-detection/android/src/main/java/expo/modules/posedetection/ExpoPoseDetectionModule.kt
package expo.modules.posedetection

import android.Manifest
import android.content.Context
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import expo.modules.interfaces.permissions.Permissions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoPoseDetectionModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoPoseDetection")

    AsyncFunction("requestCameraPermissionsAsync") { promise: Promise ->
      Permissions.askForPermissionsWithPermissionsManager(
        appContext.permissions,
        promise,
        Manifest.permission.CAMERA
      )
    }

    AsyncFunction("getCameraPermissionsAsync") { promise: Promise ->
      Permissions.getPermissionsWithPermissionsManager(
        appContext.permissions,
        promise,
        Manifest.permission.CAMERA
      )
    }

    AsyncFunction("isCameraAvailableAsync") { cameraFacing: String ->
      hasCamera(cameraFacing)
    }

    View(PoseDetectionView::class) {
      Events("onLandmarks", "onCameraReady", "onPoseError")

      Prop("active") { view: PoseDetectionView, value: Boolean ->
        view.setActiveProp(value)
      }
      Prop("cameraFacing") { view: PoseDetectionView, value: String ->
        view.setCameraFacingProp(value)
      }
      Prop("modelVariant") { view: PoseDetectionView, value: String ->
        view.setModelVariantProp(value)
      }
      Prop("minDetectionConfidence") { view: PoseDetectionView, value: Double ->
        view.setMinDetectionConfidenceProp(value.toFloat())
      }
      Prop("minTrackingConfidence") { view: PoseDetectionView, value: Double ->
        view.setMinTrackingConfidenceProp(value.toFloat())
      }
      Prop("minPresenceConfidence") { view: PoseDetectionView, value: Double ->
        view.setMinPresenceConfidenceProp(value.toFloat())
      }
      Prop("latencyDiagnosticsEnabled") { view: PoseDetectionView, value: Boolean ->
        view.setLatencyDiagnosticsEnabledProp(value)
      }
      Prop("androidPipelineMode") { view: PoseDetectionView, value: String ->
        view.setAndroidPipelineModeProp(value)
      }
      Prop("androidRotationMode") { view: PoseDetectionView, value: String ->
        view.setAndroidRotationModeProp(value)
      }
      Prop("androidAnalysisResolution") { view: PoseDetectionView, value: String ->
        view.setAndroidAnalysisResolutionProp(value)
      }
      Prop("nativeSkeletonOverlayEnabled") { view: PoseDetectionView, value: Boolean ->
        view.setNativeSkeletonOverlayEnabledProp(value)
      }
      Prop("nativeSkeletonColor") { view: PoseDetectionView, value: String ->
        view.setNativeSkeletonColorProp(value)
      }
      Prop("canvasColor") { view: PoseDetectionView, value: String ->
        view.setCanvasColorProp(value)
      }
    }
  }

  private fun hasCamera(cameraFacing: String): Boolean {
    val context = appContext.reactContext ?: return false
    val manager = context.getSystemService(Context.CAMERA_SERVICE) as? CameraManager ?: return false
    val targetLensFacing = if (cameraFacing == "back") {
      CameraCharacteristics.LENS_FACING_BACK
    } else {
      CameraCharacteristics.LENS_FACING_FRONT
    }
    return try {
      manager.cameraIdList.any { cameraId ->
        manager.getCameraCharacteristics(cameraId)
          .get(CameraCharacteristics.LENS_FACING) == targetLensFacing
      }
    } catch (_: Throwable) {
      false
    }
  }
}

File: modules/expo-pose-detection/ios/ExpoPoseDetectionModule.swift
import AVFoundation
import ExpoModulesCore

public class ExpoPoseDetectionModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoPoseDetection")

    AsyncFunction("requestCameraPermissionsAsync") { (promise: Promise) in
      switch AVCaptureDevice.authorizationStatus(for: .video) {
      case .authorized:
        promise.resolve(["status": "granted", "granted": true, "canAskAgain": true])
      case .notDetermined:
        AVCaptureDevice.requestAccess(for: .video) { granted in
          promise.resolve([
            "status": granted ? "granted" : "denied",
            "granted": granted,
            "canAskAgain": granted,
          ])
        }
      default:
        promise.resolve(["status": "denied", "granted": false, "canAskAgain": false])
      }
    }

    AsyncFunction("getCameraPermissionsAsync") { (promise: Promise) in
      switch AVCaptureDevice.authorizationStatus(for: .video) {
      case .authorized:
        promise.resolve(["status": "granted", "granted": true, "canAskAgain": true])
      case .notDetermined:
        promise.resolve(["status": "undetermined", "granted": false, "canAskAgain": true])
      default:
        promise.resolve(["status": "denied", "granted": false, "canAskAgain": false])
      }
    }

    AsyncFunction("isCameraAvailableAsync") { (cameraFacing: String) -> Bool in
      let position: AVCaptureDevice.Position = cameraFacing == "back" ? .back : .front
      let discovery = AVCaptureDevice.DiscoverySession(
        deviceTypes: [.builtInWideAngleCamera],
        mediaType: .video,
        position: position
      )
      return !discovery.devices.isEmpty
    }

    View(PoseDetectionView.self) {
      Events("onLandmarks", "onCameraReady", "onPoseError")

      Prop("active") { (view: PoseDetectionView, value: Bool) in
        view.setActiveProp(value)
      }
      Prop("cameraFacing") { (view: PoseDetectionView, value: String) in
        view.setCameraFacingProp(value)
      }
      Prop("modelVariant") { (view: PoseDetectionView, value: String) in
        view.setModelVariantProp(value)
      }
      Prop("minDetectionConfidence") { (view: PoseDetectionView, value: Double) in
        view.setMinDetectionConfidenceProp(Float(value))
      }
      Prop("minTrackingConfidence") { (view: PoseDetectionView, value: Double) in
        view.setMinTrackingConfidenceProp(Float(value))
      }
      Prop("minPresenceConfidence") { (view: PoseDetectionView, value: Double) in
        view.setMinPresenceConfidenceProp(Float(value))
      }
      Prop("latencyDiagnosticsEnabled") { (view: PoseDetectionView, value: Bool) in
        view.setLatencyDiagnosticsEnabledProp(value)
      }
    }
  }
}
